src/Components/Components/src/Microsoft.AspNetCore.Components.csproj

```diff
@@ -19,6 +23,43 @@
     <Compile Include="$(SharedSourceRoot)QueryStringEnumerable.cs" LinkBase="Shared" />
   </ItemGroup>
 
+  <ItemGroup Label="Routing">
+    <!-- Abstractions -->
+    <Compile Include="$(HttpAbstractionsSourceRoot)Routing\RouteValueDictionary.cs" LinkBase="Routing" />
+    <Compile Include="$(RoutingAbstractionsSourceRoot)IRouteConstraint.cs" LinkBase="Routing" />
+    <Compile Include="$(RoutingAbstractionsSourceRoot)IParameterPolicy.cs" LinkBase="Routing" />
+
+    <!-- Abstractions -->    
+    <Compile Include="$(SharedSourceRoot)RouteValueDictionaryTrimmerWarning.cs" LinkBase="Reflection" />
+    <Compile Include="$(SharedSourceRoot)UrlDecoder\UrlDecoder.cs" LinkBase="Routing" />
+
+    <!-- Infrastructure -->
+    <Compile Include="$(RoutingSourceRoot)RouteValueEqualityComparer.cs" LinkBase="Routing" />
+    <Compile Include="$(RoutingSourceRoot)RouteCreationException.cs" LinkBase="Routing" />
+    <!-- Tree router for matching -->
+    <Compile Include="$(RoutingSourceRoot)PathTokenizer.cs" LinkBase="Routing" />
+    <Compile Include="$(RoutingSourceRoot)Tree\*.cs" LinkBase="Routing\Tree" />
+    <Compile Remove="$(RoutingSourceRoot)Tree\Outbound*.cs" LinkBase="Routing\Tree" />
+    <Compile Remove="$(RoutingSourceRoot)Tree\LinkGeneration*.cs" LinkBase="Routing\Tree" />
+    <!-- Route patterns -->
+    <Compile Include="$(RoutingSourceRoot)Template\RoutePrecedence.cs" LinkBase="Routing\Patterns" />
+    <Compile Include="$(RoutingSourceRoot)Patterns\*.cs" LinkBase="Routing\Patterns" />
+    <Compile Remove="$(RoutingSourceRoot)Patterns\*RoutePatternTransformer.cs" />
+    <!-- Route constraints -->
+    <Compile Include="$(RoutingSourceRoot)DefaultInlineConstraintResolver.cs" LinkBase="Routing" />    
+    <Compile Include="$(RoutingSourceRoot)ParameterPolicyActivator.cs" LinkBase="Routing" />
+    <Compile Include="$(RoutingSourceRoot)RouteConstraintBuilder.cs" LinkBase="Routing" />
+    <Compile Include="$(RoutingSourceRoot)RouteConstraintMatcher.cs" LinkBase="Routing" />
+    <Compile Include="$(RoutingSourceRoot)RouteOptions.cs" LinkBase="Routing" />
+    <Compile Include="$(RoutingSourceRoot)IInlineConstraintResolver.cs" LinkBase="Routing" />
+    <Compile Include="$(RoutingSourceRoot)Constraints\**\*.cs" LinkBase="Routing\Constraints" />
+    <Compile Remove="$(RoutingSourceRoot)Constraints\HttpMethodRouteConstraint.cs" />
+    <Compile Remove="$(RoutingSourceRoot)Constraints\RegexErrorStubRouteConstraint.cs" />
+    <Compile Remove="$(RoutingSourceRoot)Constraints\RequiredRouteConstraint.cs" />
+    <Compile Remove="$(RoutingSourceRoot)Constraints\StringRouteConstraint.cs" />
```

Comment:
It's great you've got all this stuff working, and lets us have a more robust and consistent routing system.

How practical would it be to factor all this shared-source code into a library? Having so much shared-source and relying on wildcard inclusion/exclusion rules could disproportionately increase the long-term maintenance cost.

My guess is that the reasons you didn't do this are:

1. It might require a bit more factoring of implementation since you couldn't use `#if`
2. It would require us to ship a new public package (not just in the shared framework) so it can be loaded on WebAssembly

If the only concern was (1) then it might be worth pushing through with it anyway, but (2) is certainly something to take carefully.

The other main option that occurs to me is just factoring out all the `<Compile>` itemgroup entries into a separate MSBuild `.targets` file that could actually live inside `src\Http\Routing` (or the abstractions package, whatever makes most sense). At least then the responsibility for defining the set of files needed and any other conditional symbols can be factored out of `M.A.Components`, and in theory, it would even be possible to define a unit test project to cover the specific subset of routing behaviors exposed this way - that unit test project would reference the `.targets` file and thus be able to show it compiles on its own and has whatever combination of features are supposed to be made available. Having a specific unit test project for this subset would be superb, but even if there isn't time for that, just the conceptual factoring out of the dependencies into an external `.targets` file would be a cheap way of having some logical separation.

What do you think? I know we don't have unlimited time to polish everything indefinitely, but am hoping we can limit any longer-term maintenance challenge that we introduce here.

Reply:
I am not too concerned about long-term maintenance for the areas that I've touched. They don't normally receive traffic. I started the implementation by literally copying the code into M.A.C, but I chose to do it with `ifdefs` because it was a smaller change in terms of size.

For the areas where there are a lot of ifdefs, we can make the classes partial and separate them between core and server, but I chose not to do that for now to avoid more churn. (It's hard to predict ahead of time when its better to split vs when its better to ifdef).

My proposal will be that we proceed as follows:
* Apply any feedback, test, etc. we need.
* Keep files/code structure as it is for RC1 and file an issue to follow up in RC2/RTM.
* In RC2/RTM, "create a "mini-package" to more easily share routing files. Split large files with lots of ifdefs into their own.

We can do the `.targets` piece, and the unified testing project (I was already considering bringing the tests), but I suggest we do that in RC2/RTM.

There are a couple of reasons for this:
* The code for the route pattern, etc. is almost never touched (we will only touch it if we introduce new syntax, which we haven't done in years).
* The code for TreeRouter, TreeRouteBuilder, etc. is legacy from the original routing implementation, so it does not affect the current implementation for asp.net core (although there are tests to make sure that both remain 100% compatible).
  * My plan is that we start obsoleting TreeRouter and removing them, at which point they become "Blazor exclusive".

Longer term, we need to decide if we want to create a public package for routing that can be shared in some fashion, mostly to enable extensibility scenarios we currently lack.

Reply:
I've moved the definitions to a `.targets` file.